File: web/src/types/holding.ts
export interface SecurityBrief {
  id: number;
  ticker: string;
  name: string;
  exchange: string;
  asset_type: string;
}

export type ChangeType = "NEW" | "INCREASED" | "DECREASED" | "UNCHANGED" | "SOLD";

export interface HoldingSummary {
  security: SecurityBrief;
  shares: number | string;
  value_usd: number | string;
  pct_of_portfolio: number | string;
  change_type: ChangeType;
  shares_change: number | string;
  shares_change_pct: number | string | null;
}

File: backend/app/schemas/holding.py
from datetime import date
from decimal import Decimal

from pydantic import BaseModel, ConfigDict


class SecurityBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    ticker: str
    name: str
    exchange: str
    asset_type: str


class HoldingSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    security: SecurityBrief
    shares: int
    value_usd: Decimal | None = None
    pct_of_portfolio: Decimal | None = None
    change_type: str
    shares_change: int = 0
    shares_change_pct: Decimal | None = None


class HoldingHistoryPoint(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    report_date: date
    shares: int
    value_usd: Decimal | None = None
    change_type: str
